src/Generation/GapicClientGenerator.php

```diff
@@ -155,10 +156,31 @@ private function serviceName(): PhpClassMember
     private function serviceAddress(): PhpClassMember
     {
         return AST::constant('SERVICE_ADDRESS')
-            ->withPhpDocText('The default address of the service.')
+            ->withPhpDoc(PhpDoc::block(
+                PhpDoc::text("The default address of the service."),
+                $this->hasServiceAddressTemplate()
+                    ? PhpDoc::deprecated('SERVICE_ADDRESS_TEMPLATE should be used instead.')
+                    : null
+            ))
             ->withValue($this->serviceDetails->defaultHost);
     }
 
+    private function hasServiceAddressTemplate(): bool
+    {
+        return str_contains($this->serviceDetails->defaultHost, '.googleapis.com');
```

Comment:
```suggestion
        return !str_contains($this->serviceDetails->defaultHost, '.googleapis.com');
```

Reply:
I think the name implies this to be a not contains check.
Like if default host does NOT contain googleapis.com, it has a service address template.

Reply:
A service address template is the default host, but with `googleapis.com` replaced by `UNIVERSE_DOMAIN`. So if "googleapis.com" doesn't exist in the template, there's no place for us to put `UNIVERSE_DOMAIN`, so we can't give it a service address template.

Would it be more clear if we renamed the function something like `canGenerateServiceAddressTemplate`?

Reply:
my bad, this seems correct

Reply:
It may be clearer to call it `shouldHaveServiceAddressTemplate`?